tracks/haskell/exercises/pangram/mentoring.md

```diff
@@ -4,19 +4,64 @@
 import Data.Char (toLower)
 
 isPangram :: String -> Bool
-isPangram xs = all (`elem` map toLower xs) ['a'..'z']
+isPangram s = all (`elem` map toLower s) ['a'..'z']
 ```
 
+This solution traverses the input once for each letter, but terminates once each was found.
+
+It has the benefit of being declarative and to some extent short-circuiting.
+
+```haskell
+import Data.Char (toLower)
+
+isPangram :: String -> Bool
+isPangram s = null (['a'..'z'] \\ map toLower s)
+```
+
+This solution traverses the input at most once and terminates once each is found.
+
+It has the benefit of being efficient, declarative and set-based (although not `Data.Set`-based).
+
 ```haskell
 import Data.Char (toLower)
+import Data.Set (fromList, delete)
+
+alphabet :: Set Char
+alphabet = fromList ['a'..'z']
 
 isPangram :: String -> Bool
-isPangram xs = null (['a'..'z'] \\ map toLower xs)
+isPangram = any null . scanl deleteLower alphabet
+  where deleteLower cs c = delete (toLower c) cs
```

Comment:
@sshine A memory leak in this function was discovered by @mx-ws. The problem is that `delete`s bunch up behind the first `(:)` because `null` does not chase them to the end. A workaround is to use `(0 ==) . length` or `Set.delete` instead, but maybe you have other ideas?